src/sensitivities/chainrules.jl

```diff
@@ -0,0 +1,341 @@
+
+    isabstracttype(opT) || fieldcount(opT) == 0 || return false  # not handling functors
+    isempty(argTs) && return false  # we are an operator overloading AD, need operands
+
+    # Don't care about NaNMath
+    opT isa DataType && nameof(opT.name.module) == :NaNMath  && return false
+
+    # Ignore functions that have complex ranges. This may change when Nabla supports complex
+    # numbers.
+    opT ∈ typeof.((
+        SpecialFunctions.hankelh1, SpecialFunctions.hankelh2,
+        log1p, rem2pi, mod, atan, rem,
+    ))  && return false
+    opT <: Type{<:Complex} && return false  # skip complex constructor
+
+    # Ignore these functions because they have better Nabla specific versions.
+    opT ∈ typeof.((
+        isapprox, size, length, isassigned,
+        Base.Broadcast.combine_styles,  #TODO should i keep this?
+    )) && return false
+
+    # Ignore these functions because in practice they are never used and defining them cause
+    # a ton of compiler invalidations, making loading slow.
+    opT ∈ typeof.((
+        string, repr, print, println, write, readlines, eachline, Core.print, Core.println,
+        isequal, ==, in, haskey,
+        isnothing, ismissing, isfile,
+        isbitstype, isbits, isabstracttype, isconcretetype,
+        startswith, endswith, join, joinpath, normpath, chomp,
+        schedule,  # this one is huge, causes over 2500 invalidations
+    )) && return false
+
+    return true  # no exclusion applies
+end
+
+"""
+    build_def(sig)
+
+Like `ExprTools.signature` but on a signature type-tuple, not a Method.
+For `sig` being a tuple-type representing a methods type signature, this generates a
+dictionary that can be passes to `ExprTools.combinedef` to define that function,
+Provided that you assign the `:body` key on the dictionary first.
+
+For example:
+```julia
+julia> Nabla.build_def(Tuple{typeof(identity), Any})
+Dict{Symbol, Any} with 2 entries:
+  :name => :(op::typeof(identity))
+  :args => Expr[:(x1::Any)]
+
+julia> Nabla.build_def(Tuple{typeof(+), Vector{T}, Vector{T}} where T<:Number)
+Dict{Symbol, Any} with 3 entries:
+  :name        => :(op::typeof(+))
+  :args        => Expr[:(x1::Array{var"##T#5492", 1}), :(x2::Array{var"##T#5492", 1})]
+  :whereparams => Any[:(var"##T#5492" <: Number)]
+```
+"""
+function build_def(orig_sig)
+    sig = _truly_rename_unionall(orig_sig)  # TODO ExprTools possibly should do this for `signature(::Method)`` also
+    def = Dict{Symbol, Any}()
+
+    opT = ExprTools.parameters(sig)[1]
+    def[:name] = :(op::$opT)
+
+    arg_types = ExprTools.name_of_type.(ExprTools.argument_types(sig))
+    arg_names = [Symbol(:x, ii) for ii in eachindex(arg_types)]  #TODO: should we pass the arg_names in?
+    def[:args] = Expr.(:(::), arg_names, arg_types)
+    def[:whereparams] = ExprTools.where_parameters(sig)
+
+    filter!(kv->last(kv)!==nothing, def)  # filter out nonfields.
+    return def
+end
```

Comment:
Seems reasonable to move this to ExprTools. Should probably be renamed to `signature` though

Reply:
done